// @nodejs/spider-common/lib/constants.js
/**
 * @author yangyufei
 * @date 2019-02-26 13:27:51
 * @desc
 */
const CONSTANTS = module.exports = {
    ENV     : process.env.NODE_ENV || 'development',
    MONITOR : true, // 默认加入监控当中

    // 错误相关信息
    ERROR_OBJ   : {
        SUCCESS     : {code: 0, msg: '操作成功！'},

        DEFAULT     : {code: 100, msg: '系统错误！'},
        TIMEOUT     : {code: 101, msg: '请求访问超时！'},
        RETRYOUT    : {code: 102, msg: '超过最大重试次数！'},
        PARSEJSON   : {code: 103, msg: '异常非json数据！'},
        BAD_REQUEST : {code: 104, msg: 'uri请求错误！'},
        BAD_CONFIG  : {code: 105, msg: '配置错误！'},
        CHECK_RULE  : {code: 106, msg: '网站接口/页面规则校验不通过！'},
        BAD_OUTPUT  : {code: 107, msg: '输出数据校验失败！'}
    },

    // 网络监控相关keys
    NET_MONITOR_KEYS: {
        STATE_NET   : 'network:connect:state', // 当前网络基本状态
        NET_LAST_TEST: 'network:connect:lastTestTime', // 上次检查网络状态时间
        POOL        : 'network:proxy:pool', // 代理池
    },

    // 网络状态
    NET_STATE       : {
        DISCONNECT  : -1, // 网络不通
        GOOD        : 1, // 通畅
    },

    TASK_STATUS     : {
        BIG_RECORD  : -2, // 查询条件下数据过多，需要再次分割
        ERROR       : -1, // 失败
        WAITING     : 0, // 等待
        RUNNING     : 1, // 运行中
        SUCCESS     : 2, // 成功
    },

    OUT_FILE_SIZE   : 500, // 输出out文件的大小

    // 过滤器类型
    FILTER_TYPE     : {
        SIMPLE      : 1, // 简单过滤器
        EXPIRE      : 2, // 带过期的过滤器
    },

    // 爬虫相关的默认配置
    spider          : {
        // 请求接口相关配置项
        fetch   : {
            timeout     : 20000, // http访问超时时间
            retry       : 5, // 访问错误重试次数
        },

        // 任务执行选项
        task    : {
            concurrency : 100, // 并发数据
            retry       : 5, // 任务错误重试次数
        },

        /**
         * 运行方式，共3种，分别是
         * once     ：运行一次，用于抓取历史累积数据；
         * forever  ：一直运行，用于抓取无法估计每轮更新时间的任务；
         * cron     ：定时任务，一般是每日更新
         */
        run     : {
            type: 'once', // 运行方式，共3种：once
        },
    },

    // redis配置
    redis           : {
        host    : '127.0.0.1',
        port    : 6379,
    }
};
